packages/consensus/src/types/utreexo.cairo

```diff
@@ -72,8 +74,74 @@ pub trait UtreexoAccumulator {
     fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof);
 }
 
+pub impl UtreexoStateImpl of UtreexoAccumulator {
+    /// Adds single output to the accumulator.
+    /// The order *is important*: adding A,B and B,A would result in different states.
+    ///
+    /// Note that this call also pushes old UTXOs "to the left", to a larger subtree.
+    /// This mechanism ensures that short-lived outputs have small inclusion proofs.
+    fn add(ref self: UtreexoState, output: OutPoint) {}
+
+    /// Verifies inclusion proof for a single output.
+    fn verify(
+        self: @UtreexoState, output: @OutPoint, proof: @UtreexoProof
+    ) -> Result<(), UtreexoError> {
+        let proof_root = verify_helper(*proof.proof, *proof.leaf_index, (*output.txid).into());
+
+        // Get the expected root from the last node in the proof path.
+        let expected_root: u256 = (*proof.proof[(*proof.proof).len() - 1]).into();
+
+        if expected_root == proof_root {
+            Result::Ok(())
+        } else {
+            Result::Err(UtreexoError::ProofVerificationFailed)
+        }
+    }
+
+
+    /// Removes single output from the accumlator (order is important).
+    ///
+    /// Note that once verified, the output itself is not required for deletion,
+    /// the leaf index plus inclusion proof is enough.
+    fn delete(ref self: UtreexoState, proof: @UtreexoProof) {}
+
+    /// Verifies batch proof for multiple outputs (e.g. all outputs in a block).
+    fn verify_batch(
+        self: @UtreexoState, outputs: Span<OutPoint>, proof: @UtreexoBatchProof
+    ) -> Result<(), UtreexoError> {
+        Result::Ok(())
+    }
+
+    /// Removes multiple outputs from the accumulator.
+    fn delete_batch(ref self: UtreexoState, proof: @UtreexoBatchProof) {}
+}
+
+fn verify_helper(mut proof: Span<felt252>, leaf_index: u64, curr_node: u256) -> u256 {
+    if proof.len() == 0 {
+        return curr_node;
+    }
+    let next_index = (leaf_index - leaf_index % 2) / 2;
```

Comment:
This is the same is `leaf_index / 2`, you can do `next_index, is_right =  DivRem::div_rem(leaf_index, 2)`.